.travis.yml

```diff
@@ -1,62 +1,71 @@
-language: c
+language: minimal
 sudo: false
 notifications:
   email:
     - tbekolay@gmail.com
+cache:
+  directories:
+  - $HOME/miniconda
 
 env:
   global:
+    - MODE="test"
     - NUMPY="1.14"
     - SCIPY="false"
     - COVERAGE="false"
-    - STATIC="false"
-    - DOCS="false"
 
-matrix:
+jobs:
   include:
-    - env: PYTHON="2.7" STATIC="true"
-    - env: PYTHON="2.7" COVERAGE="true"
-    - env: PYTHON="2.7" NUMPY="1.11"
-    - env: PYTHON="2.7" NUMPY="1.12"
-    - env: PYTHON="2.7" NUMPY="1.13"
-    - env: PYTHON="2.7" NUMPY="1.14"
-    - env: PYTHON="3.4" NUMPY="1.11"  # <=1.12 not packaged for py34 in conda
-    - env: PYTHON="3.5"
-    - env: PYTHON="3.6" COVERAGE="true" SCIPY="true"
-    - env: PYTHON="3.6" DOCS="true"
+    - stage: basic
+      env: MODE="static" PYTHON="2.7"
+    - env: MODE="test" PYTHON="2.7" COVERAGE="true"
+    - stage: advanced
+      env: MODE="test" PYTHON="2.7" NUMPY="1.11"
+      cache: false  # disable the cache for one build to make sure that works
+    - env: MODE="test" PYTHON="2.7" NUMPY="1.12"
+    - env: MODE="test" PYTHON="2.7" NUMPY="1.13"
+    - env: MODE="test" PYTHON="2.7" NUMPY="1.14"
+    # <=1.12 not packaged for py34 in conda
+    - env: MODE="test" PYTHON="3.4" NUMPY="1.11"
+    - env: MODE="test" PYTHON="3.5"
+    - env: MODE="test" PYTHON="3.6" COVERAGE="true" SCIPY="true"
+    - env: MODE="docs" PYTHON="3.6"
+    - stage: deploy
+      if: branch =~ ^release-candidate-* OR tag =~ ^v[0-9]*
+      env: MODE="deploy" PYTHON="3.6"
+      deploy:
+        - provider: pypi
+          server: https://test.pypi.org/legacy/
+          user: tbekolay
+          password: $PYPI_TOKEN
+          distributions: "sdist bdist_wheel"
+          on:
+            all_branches: true
+            condition: $TRAVIS_BRANCH =~ ^release-candidate-*
+            condition: $TRAVIS_TAG = ""
+        - provider: pypi
+          user: tbekolay
+          password: $PYPI_TOKEN
+          distributions: "sdist bdist_wheel"
+          on:
+            all_branches: true
+            condition: $TRAVIS_TAG =~ ^v[0-9]*
 
 before_install:
+  - chmod +x .ci/deploy.sh  # so we can keep it -x on local machines
```

Comment:
Why do we want to keep this `-x` on local machines?

Reply:
So that somebody doesn't call it accidentally, I assume. This is the way it was before; I just added the comment. I had actually switched it, then I thought that this was probably done intentionally, so I added the comment instead.

Reply:
I think it was initially because marking things as executable is hard (impossible?) on Windows, but I think we should just mark it executable.